### PR TITLE
Fixes #1809 - Can now remove a collection while filters are being loa…

### LINKED ISSUE #1809: Removing a collection before it has finished loading results in errors
**Steps to reproduce**
Add a collection and then remove it again before it has finished loading

**What should happen**
Collection should be removed

**What does happen**
Many JavaScript errors

### PR DESCRIPTION
**Pull request by jonescc**
…ded without errors.

### CODE REVIEW

**Comment by danfruehauf** [comment]
Can still reproduce this on this branch:
```
TypeError: this.layer is null
...eturn b},shouldDraw:function(){var a=!1,b=this.layer.maxExtent;if(b){var c=this....
```

File: src/test/javascript/portal/prototypes/Ext.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+            testComponent = new Ext.Component();
+            callbackSpy = jasmine.createSpy('callbackSpy');
+        });
+        
+        it("calls callback on component if component hasn't been destroyed", function() {
+            var testCallback = testComponent.createSafeCallback(callbackSpy);
+            testCallback("test");
+            expect(callbackSpy).toHaveBeenCalledWith("test");
+        });
+
```
Comment (danfruehauf) [inline src/test/javascript/portal/prototypes/Ext.js]: Empty line with spaces.

File: src/test/javascript/portal/prototypes/Ext.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+describe("Ext.Component", function() {
+    describe('createSafeCallback', function() {
+        var testComponent;
+        var callbackSpy;
+        
+        beforeEach(function() {
+            testComponent = new Ext.Component();
+            callbackSpy = jasmine.createSpy('callbackSpy');
+        });
+
```
Comment (danfruehauf) [inline src/test/javascript/portal/prototypes/Ext.js]: Here too...

File: src/test/javascript/portal/prototypes/Ext.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+ * Copyright 2014 IMOS
+ *
+ * The AODN/IMOS Portal is distributed under the terms of the GNU General Public License
+ *
+ */
+describe("Ext.Component", function() {
+    describe('createSafeCallback', function() {
+        var testComponent;
+        var callbackSpy;
+
```
Comment (danfruehauf) [inline src/test/javascript/portal/prototypes/Ext.js]: And here.

**Comment by jonescc** [comment]
Changes made @danfruehauf

File: web-app/js/portal/prototypes/OpenLayers.js [hunk 4]
```diff
@@ -258,3 +258,26 @@ OpenLayers.Geometry.prototype.crossesAntimeridian = function() {
...
+    if (this.destroyDeferred) {
+        return;
+    }
+
+    this.events.register('loadend', this, function() {
+        this.destroy();
+    });
+
+    this.destroyDeferred = true;
+}
```
Comment (danfruehauf) [inline web-app/js/portal/prototypes/OpenLayers.js]: I think missing semicolon here.

File: web-app/js/portal/prototypes/Ext.js [hunk 5]
```diff
@@ -47,3 +49,19 @@ Ext.tree.TreeNode.prototype.getHierarchy = function(attribute) {
...
+
+Ext.Component.prototype.createSafeCallback = function(callback) {
+    var callingComponent = this;
+    
+    return function() {
+        if (typeof callingComponent.isDestroyed === "undefined" || !callingComponent.isDestroyed) {
+            callback.apply(this, arguments);
+        }
+    }
+}
```
Comment (danfruehauf) [inline web-app/js/portal/prototypes/Ext.js]: Missing semicolon.

**Comment by danfruehauf** [comment]
Fix works. Just the couple semicolons.

File: src/test/javascript/portal/prototypes/OpenLayersSpec.js [hunk 6]
```diff
@@ -228,4 +228,30 @@ describe('OpenLayers', function() {
             expect(window.trackUsage).toHaveBeenCalledWith('Filters', 'Spatial Constraint', 'type=a style  that the current style is not set to', undefined);
         });
     });
+
```
Comment (danfruehauf) [inline src/test/javascript/portal/prototypes/OpenLayersSpec.js]: Empty line :)

File: web-app/js/portal/filter/ui/FilterGroupPanel.js [hunk 7]
```diff
@@ -272,4 +276,5 @@ Portal.filter.ui.FilterGroupPanel = Ext.extend(Ext.Container, {
     _isGlobalFilterPanel: function(panel) {
         return panel.filter.type == "geometrypropertytype";
     }
+
```
Comment (danfruehauf) [inline web-app/js/portal/filter/ui/FilterGroupPanel.js]: And here, just spaces.

File: web-app/js/portal/prototypes/Ext.js [hunk 8]
```diff
@@ -47,3 +49,19 @@ Ext.tree.TreeNode.prototype.getHierarchy = function(attribute) {
...
 };
+
+// Ext.Component additions
+
+// Function to wrap a supplied callback so that it will not be called if the calling
+// component has been destroyed.
+
+Ext.Component.prototype.createSafeCallback = function(callback) {
+    var callingComponent = this;
+
```
Comment (danfruehauf) [inline web-app/js/portal/prototypes/Ext.js]: Here too. I think your editor needs some new settings...

**Comment by danfruehauf** [comment]
@jonescc I amended this commit. No need to do anything further.

**Pull request merged**